### PR TITLE
[NPU] Support npu op `expand`

### PR DESCRIPTION
**Pull request by Meiyim**
### PR types
New features
### PR changes
OPs

### Describe
+ ascend `expand` aka `Tile`

image

c++ test
image

python test

infoflow 2021-03-15 22-44-45

infoflow 2021-03-15 22-46-03

### CODE REVIEW

File: paddle/fluid/operators/expand_op.h [hunk 1]
```diff
@@ -48,14 +48,23 @@ namespace paddle {
 namespace operators {
 inline std::vector<int> get_expand_times(
     const framework::ExecutionContext& ctx) {
+  std::cout << "get_expand_times" << std::endl;
```
Comment (liym27) [inline paddle/fluid/operators/expand_op.h]: Delete useless code

File: paddle/fluid/operators/expand_op.h [hunk 2]
```diff
@@ -48,14 +48,23 @@ namespace paddle {
...
   if (ctx.HasInput("ExpandTimes")) {
+    std::cout << "get_expand_times2" << std::endl;
```
Comment (liym27) [inline paddle/fluid/operators/expand_op.h]: Delete useless code

File: paddle/fluid/operators/expand_op.h [hunk 3]
```diff
@@ -48,14 +48,23 @@ namespace paddle {
...
     auto* expand_tensor = ctx.Input<framework::LoDTensor>("ExpandTimes");
     auto* expand_data = expand_tensor->data<int>();
     framework::Tensor cpu_expand_tensor;
     if (platform::is_gpu_place(expand_tensor->place())) {
       TensorCopySync(*expand_tensor, platform::CPUPlace(), &cpu_expand_tensor);
       expand_data = cpu_expand_tensor.data<int>();
     }
+#ifdef PADDLE_WITH_ASCEND_CL
+    if (platform::is_npu_place(expand_tensor->place())) {
+      std::cout << "get_expand_times3" << std::endl;
```
Comment (liym27) [inline paddle/fluid/operators/expand_op.h]: The same as above

**Review by liym27 — COMMENTED** [review]
Should we support `expand` npu kernel not `expand_v2 `？  expand_v2 is the latest version, the old version  `expand` will be gradually deprecated since Paddle 2.0.0

File: paddle/fluid/operators/expand_op_npu.cc [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+namespace ops = paddle::operators;
+REGISTER_OP_NPU_KERNEL(
+    expand, ops::ExpandNPUKernel<paddle::platform::NPUDeviceContext, float>,
+    ops::ExpandNPUKernel<paddle::platform::NPUDeviceContext,
+                         paddle::platform::float16>,
+    ops::ExpandNPUKernel<paddle::platform::NPUDeviceContext, int>,
+    ops::ExpandNPUKernel<paddle::platform::NPUDeviceContext, uint8_t>,
+    ops::ExpandNPUKernel<paddle::platform::NPUDeviceContext, int8_t>);
+
```
Comment (liym27) [inline paddle/fluid/operators/expand_op_npu.cc:81]: Add unittests for float16, int, uint8_t, int64_t

File: paddle/fluid/operators/expand_op_npu_test.cc [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  auto place = ctx.GetPlace();
+  TensorFromVector(std::vector<T>(3 * 1 * 7, 1), ctx, in_t);
+  TensorFromVector(std::vector<int>({1, 10, 1}), ctx, expand_times_t);
+
+  in_t->Resize(f::make_ddim({3, 1, 7}));
+  expand_times_t->Resize(f::make_ddim({3}));
+  out_t->Resize(f::make_ddim({3, 10, 7}));
+
+  // in_t->mutable_data<T>(place);
+  // expand_times_t->mutable_data<int>(place);
```
Comment (liym27) [inline paddle/fluid/operators/expand_op_npu_test.cc]: Delete useless code

File: python/paddle/fluid/tests/unittests/npu/test_expand_op_npu.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+        a_np = np.random.random(size=(32, 1)).astype('float32')
+        label_np = np.random.randint(2, size=(32, 1)).astype('int64')
+
+        with paddle.static.program_guard(main_prog, startup_prog):
+            a = paddle.static.data(name="a", shape=[32, 1], dtype='float32')
+            label = paddle.static.data(
+                name="label", shape=[32, 1], dtype='int64')
+
+            res = paddle.fluid.layers.expand(a, [1,32])
```
Comment (liym27) [inline python/paddle/fluid/tests/unittests/npu/test_expand_op_npu.py:107]: why not call `paddle.expand` ?

**Comment by liym27** [comment]
Please specify PR Title Description. 

For example:
[NPU] Support npu op gelu and gelu_grad

**Comment by liym27** [comment]
Please specify PR Title Description.

For example:
[NPU] Support npu op gelu and gelu_grad

File: paddle/fluid/operators/expand_op.h (hunk 1 above)
Comment (Meiyim) [reply paddle/fluid/operators/expand_op.h]: removed

File: paddle/fluid/operators/expand_op.h (hunk 2 above)
Comment (Meiyim) [reply paddle/fluid/operators/expand_op.h]: removed

File: paddle/fluid/operators/expand_op.h (hunk 3 above)
Comment (Meiyim) [reply paddle/fluid/operators/expand_op.h]: removed

File: paddle/fluid/operators/expand_op_npu_test.cc (hunk 5 above)
Comment (Meiyim) [reply paddle/fluid/operators/expand_op_npu_test.cc]: removed

File: python/paddle/fluid/tests/unittests/npu/test_expand_op_npu.py (hunk 6 above)
Comment (Meiyim) [reply python/paddle/fluid/tests/unittests/npu/test_expand_op_npu.py:107]: paddle.expand use `expand_v2`. unittest intent to use `expand`

File: python/paddle/fluid/tests/unittests/npu/test_expand_op_npu.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        self.set_npu()
+        self.op_type = "expand"
+        self.place = paddle.NPUPlace(0)
+
+        self.init_dtype()
+        np.random.seed(SEED)
+        x = np.random.randn(3,1,7).astype(self.dtype)
+        out = np.tile(x, [1,10,1])
+
+        self.inputs = {'X': OpTest.np_dtype_to_fluid_dtype(x)}
```
Comment (pangyoki) [inline python/paddle/fluid/tests/unittests/npu/test_expand_op_npu.py:42]: 新增一个单测样例，需要验证输入Tensor为`ExpandTimes`的情况。

**Review by pangyoki — COMMENTED** [review]
截图中没有反向`expand_grad`的信息，也没有添加NPU的expand反向kernel。请关注下反向的情况。

File: paddle/fluid/operators/expand_op_npu.cc (hunk 4 above)
Comment (Meiyim) [reply paddle/fluid/operators/expand_op_npu.cc:81]: removed int* kernel for now

File: python/paddle/fluid/tests/unittests/npu/test_expand_op_npu.py (hunk 7 above)
Comment (Meiyim) [reply python/paddle/fluid/tests/unittests/npu/test_expand_op_npu.py:42]: 已经更新

**Comment by Meiyim** [comment]
`expand_grad` 在当前组网中没有使用，可以先低优考虑

**Comment by liym27** [comment]
Should we support expand npu kernel not expand_v2 ？ expand_v2 is the latest version, the old version expand will be gradually deprecated since Paddle 2.0.0？

**Comment by Meiyim** [comment]
当前组网采用的是`expand`实现，暂时无需考虑`expand_v2`。

**Review by pangyoki — COMMENTED** [review]
LGTM

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**